src/cfg.rs

```diff
@@ -415,124 +400,123 @@ impl SExpr {
-            while paren_stack_size > 0 {
-                let token = match tokens.next() {
-                    None => bail!(
-                        "Sub expression does not close:\n{}\nwhole expr:\n{}",
-                        subexpr,
-                        expr
-                    ),
-                    Some(t) => t,
-                };
-                paren_stack_size += token.chars().filter(|c| *c == '(').count();
-                paren_stack_size -= token.chars().filter(|c| *c == ')').count();
-                subexpr.push(' ');
-                subexpr.push_str(token);
-            }
-            ret.push(SExpr::List(parse_expr(&subexpr)?))
-        } else if token.contains(')') {
-            bail!(
-                "Unexpected closing paren in token {} in expr:\n{}",
-                token,
-                expr
-            )
-        } else if token.starts_with('"') {
-            let mut token = token.strip_prefix('"').unwrap();
-            let mut quoted_tokens = vec![];
-            loop {
-                let num_dquotes = token.matches('"').count();
-                // seek to end of quoted string
-                match num_dquotes {
-                    0 => {
-                        quoted_tokens.push(token);
-                        token = match tokens.next() {
-                            Some(t) => t,
-                            None => {
-                                bail!("Unterminated quoted string: {}", quoted_tokens.join(" "))
-                            }
-                        };
+            Some(t) => match t? {
+                Open => stack.push(vec![]),
+                Close => {
+                    let expr = List(stack.pop().unwrap());
+                    if stack.is_empty() {
+                        bail!("Unexpected paren");
                     }
-                    1 => {
-                        if !token.ends_with('"') {
-                            bail!("Invalid end of quoted string {}", token);
-                        }
-                        quoted_tokens.push(token.strip_suffix('"').unwrap());
-                        break;
-                    }
-                    _ => bail!(
-                        "Invalid quoted string \"{} {}",
-                        quoted_tokens.join(" "),
-                        token
-                    ),
+                    stack.last_mut().unwrap().push(expr);
                 }
-            }
-            ret.push(SExpr::Atom(quoted_tokens.join(" ")));
-        } else if token.contains('"') {
-            // token contains " but does not start with "; not valid
-            bail!("Invalid start of quoted string: {}", token);
-        } else {
-            ret.push(SExpr::Atom(token.to_owned()));
+                String(s) => stack.last_mut().unwrap().push(Atom(s.to_string())),
+            },
         }
     }
-    Ok(ret)
+    let exprs = stack.pop().unwrap();
+    if !stack.is_empty() {
+        bail!("Unclosed parens");
```

Comment:
It would improve the error message to include the text involved with the unclosed parenthesis here. It might potentially be a lot of text, so could consider truncating to a certain number of lines/characters, but not necessary.